internal/webhooks/handler.go

```diff
@@ -0,0 +1,191 @@
+	envSplunkOtelAgent          = "SPLUNK_OTEL_AGENT"
+	envOTELServiceName          = "OTEL_SERVICE_NAME"
+	envOTELExporterOTLPEndpoint = "OTEL_EXPORTER_OTLP_ENDPOINT"
+	envOTELTracesExporter       = "OTEL_TRACES_EXPORTER"
+	envOTELResourceAttrs        = "OTEL_RESOURCE_ATTRIBUTES"
+
+	exporterOTLP   = "otlp"
+	exporterJaeger = "jaeger-thrift-splunk"
+
+	annotationConfInjectionEnabled = "o11y.splunk.com/inject-config"
+)
+
+type handler struct {
+	client  client.Client
+	logger  logr.Logger
+	decoder *admission.Decoder
+}
+
+type config struct {
+	exporter string
+	endpoint string
+}
+
+// NewHandler creates a new WebhookHandler.
+func NewHandler(logger logr.Logger, cl client.Client) admission.Handler {
+	return &handler{
+		client: cl,
+		logger: logger,
+	}
+}
+
+func (h *handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	pod := corev1.Pod{}
+	err := h.decoder.Decode(req, &pod)
+	if err != nil {
+		h.logger.Error(err, "unable to decode pod")
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	if !strings.EqualFold(pod.Annotations[annotationConfInjectionEnabled], "true") {
+		return admission.Allowed("")
+	}
+
+	// we use the req.Namespace here because the pod might have not been created yet
+	ns := corev1.Namespace{}
+	err = h.client.Get(ctx, types.NamespacedName{Name: req.Namespace, Namespace: ""}, &ns)
+	if err != nil {
+		h.logger.Error(err, "unable to get pod namespace", "namespace", req.Namespace)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	spec, err := h.getAgentSpec(ctx)
+	if err != nil {
+		h.logger.Error(err, "unable to get splunk agent spec. make sure SpluknOtelAgent is deployed")
+		return admission.Allowed("")
+	}
+
+	cfg := configFromSpec(spec)
+
+	pod = h.injectConfigIntoPod(ctx, cfg, pod, ns)
+	marshaledPod, err := json.Marshal(pod)
+	if err != nil {
+		h.logger.Error(err, "unable to marshal pod", "pod", pod.Name)
+		return admission.Allowed("")
+	}
+
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledPod)
+}
+
+func configFromSpec(spec *v1alpha1.SplunkOtelAgentSpec) config {
+
+	cfg := config{
+		exporter: exporterOTLP,
+		endpoint: "http://$(SPLUNK_OTEL_AGENT):4317",
+	}
+
+	if spec.Agent.Disabled {
+		if !spec.Gateway.Disabled {
+			cfg.endpoint = fmt.Sprintf("http://splunk-otel-collector.%s:4317", operatorNamespace)
+		} else {
+			cfg.exporter = exporterJaeger
+			cfg.endpoint = fmt.Sprintf("https://ingest.%s.signalfx.com/v2/trace", spec.Realm)
+		}
+	}
+
+	return cfg
+}
+
+func (h *handler) injectConfigIntoPod(ctx context.Context, cfg config, pod corev1.Pod, ns corev1.Namespace) corev1.Pod {
+	if len(pod.Spec.Containers) < 1 {
+		h.logger.Info("no containers found in pod", "pod", pod.Name)
+		return pod
+	}
+
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	container := &pod.Spec.Containers[0]
+	attrs := map[string]string{}
+	resourceAttrs := h.createResourceMap(ctx, attrs, ns, pod)
```

Comment:
Looks like `createResourceMap` gathers more attributes than is needed for `serviceName` call. Why do we need them? 

Maybe we can just fetch all required info right from `serviceName` instead of capturing all the data (with some redundant fields?) in advance?

Reply:
This is used to populate OTEL_RESOURCE_ATTRIBUTES env var but that got left out. Will be added in a follow up PR

Reply:
Added relevant code